File: binding/src/libqitemeditorfactory.cpp
#include <QByteArray>
#include <QItemEditorCreatorBase>
#include <QItemEditorFactory>
#include <QWidget>
#include <qitemeditorfactory.h>
#include "libqitemeditorfactory.h"
#include "libqitemeditorfactory.hxx"

QItemEditorCreatorBase* QItemEditorCreatorBase_new() {
	 return new VirtualQItemEditorCreatorBase();
}

QItemEditorCreatorBase* QItemEditorCreatorBase_new2(const QItemEditorCreatorBase* param1) {
	 return new VirtualQItemEditorCreatorBase(*param1);
}

QWidget* QItemEditorCreatorBase_CreateWidget(const QItemEditorCreatorBase* self, QWidget* parent) {
	return self->createWidget(parent);
}

libqt_string QItemEditorCreatorBase_ValuePropertyName(const QItemEditorCreatorBase* self) {
	QByteArray _qb = self->valuePropertyName();
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QItemEditorCreatorBase_OperatorAssign(QItemEditorCreatorBase* self, const QItemEditorCreatorBase* param1) {
	self->operator=(*param1);
}

// Base class handler implementation
QWidget* QItemEditorCreatorBase_QBaseCreateWidget(const QItemEditorCreatorBase* self, QWidget* parent) {
	auto* vqitemeditorcreatorbase = dynamic_cast<const VirtualQItemEditorCreatorBase*>(self);
	if (vqitemeditorcreatorbase && vqitemeditorcreatorbase->isVirtualQItemEditorCreatorBase) {
vqitemeditorcreatorbase->setQItemEditorCreatorBase_CreateWidget_IsBase(true);
	return vqitemeditorcreatorbase->createWidget(parent);
}
}

// Auxiliary method to allow providing re-implementation
void QItemEditorCreatorBase_OnCreateWidget(const QItemEditorCreatorBase* self, intptr_t slot) {
	auto* vqitemeditorcreatorbase = dynamic_cast<const VirtualQItemEditorCreatorBase*>(self);
	if (vqitemeditorcreatorbase && vqitemeditorcreatorbase->isVirtualQItemEditorCreatorBase) {
vqitemeditorcreatorbase->setQItemEditorCreatorBase_CreateWidget_Callback(reinterpret_cast<VirtualQItemEditorCreatorBase::QItemEditorCreatorBase_CreateWidget_Callback>(slot));
}
}

// Base class handler implementation
libqt_string QItemEditorCreatorBase_QBaseValuePropertyName(const QItemEditorCreatorBase* self) {
	auto* vqitemeditorcreatorbase = dynamic_cast<const VirtualQItemEditorCreatorBase*>(self);
	if (vqitemeditorcreatorbase && vqitemeditorcreatorbase->isVirtualQItemEditorCreatorBase) {
vqitemeditorcreatorbase->setQItemEditorCreatorBase_ValuePropertyName_IsBase(true);
	QByteArray _qb = vqitemeditorcreatorbase->valuePropertyName();
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}
}

// Auxiliary method to allow providing re-implementation
void QItemEditorCreatorBase_OnValuePropertyName(const QItemEditorCreatorBase* self, intptr_t slot) {
	auto* vqitemeditorcreatorbase = dynamic_cast<const VirtualQItemEditorCreatorBase*>(self);
	if (vqitemeditorcreatorbase && vqitemeditorcreatorbase->isVirtualQItemEditorCreatorBase) {
vqitemeditorcreatorbase->setQItemEditorCreatorBase_ValuePropertyName_Callback(reinterpret_cast<VirtualQItemEditorCreatorBase::QItemEditorCreatorBase_ValuePropertyName_Callback>(slot));
}
}

void QItemEditorCreatorBase_Delete(QItemEditorCreatorBase* self) {
    delete self;
}

QItemEditorFactory* QItemEditorFactory_new() {
	 return new VirtualQItemEditorFactory();
}

QItemEditorFactory* QItemEditorFactory_new2(const QItemEditorFactory* param1) {
	 return new VirtualQItemEditorFactory(*param1);
}

QWidget* QItemEditorFactory_CreateEditor(const QItemEditorFactory* self, int userType, QWidget* parent) {
	return self->createEditor(userType, parent);
}

libqt_string QItemEditorFactory_ValuePropertyName(const QItemEditorFactory* self, int userType) {
	QByteArray _qb = self->valuePropertyName(userType);
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QItemEditorFactory_RegisterEditor(QItemEditorFactory* self, int userType, QItemEditorCreatorBase* creator) {
	self->registerEditor(userType, creator);
}

const QItemEditorFactory* QItemEditorFactory_DefaultFactory() {
	return QItemEditorFactory::defaultFactory();
}

void QItemEditorFactory_SetDefaultFactory(QItemEditorFactory* factory) {
	QItemEditorFactory::setDefaultFactory(factory);
}

// Base class handler implementation
QWidget* QItemEditorFactory_QBaseCreateEditor(const QItemEditorFactory* self, int userType, QWidget* parent) {
	auto* vqitemeditorfactory = dynamic_cast<const VirtualQItemEditorFactory*>(self);
	if (vqitemeditorfactory && vqitemeditorfactory->isVirtualQItemEditorFactory) {
vqitemeditorfactory->setQItemEditorFactory_CreateEditor_IsBase(true);
	return vqitemeditorfactory->createEditor(userType, parent);
}
}

// Auxiliary method to allow providing re-implementation
void QItemEditorFactory_OnCreateEditor(const QItemEditorFactory* self, intptr_t slot) {
	auto* vqitemeditorfactory = dynamic_cast<const VirtualQItemEditorFactory*>(self);
	if (vqitemeditorfactory && vqitemeditorfactory->isVirtualQItemEditorFactory) {
vqitemeditorfactory->setQItemEditorFactory_CreateEditor_Callback(reinterpret_cast<VirtualQItemEditorFactory::QItemEditorFactory_CreateEditor_Callback>(slot));
}
}

// Base class handler implementation
libqt_string QItemEditorFactory_QBaseValuePropertyName(const QItemEditorFactory* self, int userType) {
	auto* vqitemeditorfactory = dynamic_cast<const VirtualQItemEditorFactory*>(self);
	if (vqitemeditorfactory && vqitemeditorfactory->isVirtualQItemEditorFactory) {
vqitemeditorfactory->setQItemEditorFactory_ValuePropertyName_IsBase(true);
	QByteArray _qb = vqitemeditorfactory->valuePropertyName(userType);
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}
}

// Auxiliary method to allow providing re-implementation
void QItemEditorFactory_OnValuePropertyName(const QItemEditorFactory* self, intptr_t slot) {
	auto* vqitemeditorfactory = dynamic_cast<const VirtualQItemEditorFactory*>(self);
	if (vqitemeditorfactory && vqitemeditorfactory->isVirtualQItemEditorFactory) {
vqitemeditorfactory->setQItemEditorFactory_ValuePropertyName_Callback(reinterpret_cast<VirtualQItemEditorFactory::QItemEditorFactory_ValuePropertyName_Callback>(slot));
}
}

void QItemEditorFactory_Delete(QItemEditorFactory* self) {
    delete self;
}

File: binding/include/libqitemeditorfactory.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qitemeditorcreatorbase.html
pub const qitemeditorcreatorbase = struct {

    /// New constructs a new QItemEditorCreatorBase object.
    pub fn New() QtC.QItemEditorCreatorBase {
        return qtc.QItemEditorCreatorBase_new();
    }


    /// New2 constructs a new QItemEditorCreatorBase object.
    pub fn New2(param1: ?*anyopaque) QtC.QItemEditorCreatorBase {
        return qtc.QItemEditorCreatorBase_new2(@ptrCast(param1));
    }


    pub fn CreateWidget(self: ?*anyopaque, parent: ?*anyopaque) QtC.QWidget {
        return qtc.QItemEditorCreatorBase_CreateWidget(@ptrCast(self), @ptrCast(parent));
    }

    /// Allows for overriding the related default method
    pub fn OnCreateWidget(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) QtC.QWidget) void {
        qtc.QItemEditorCreatorBase_OnCreateWidget(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseCreateWidget(self: ?*anyopaque, parent: ?*anyopaque) QtC.QWidget {
        return qtc.QItemEditorCreatorBase_QBaseCreateWidget(@ptrCast(self), @ptrCast(parent));
    }

    pub fn ValuePropertyName(self: ?*anyopaque, , allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QItemEditorCreatorBase_ValuePropertyName(@ptrCast(self));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qitemeditorcreatorbase.ValuePropertyName: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    /// Allows for overriding the related default method
    pub fn OnValuePropertyName(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) [*:0]u8) void {
        qtc.QItemEditorCreatorBase_OnValuePropertyName(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseValuePropertyName(self: ?*anyopaque, , allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QItemEditorCreatorBase_QBaseValuePropertyName(@ptrCast(self));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qitemeditorcreatorbase.ValuePropertyName: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn OperatorAssign(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QItemEditorCreatorBase_OperatorAssign(@ptrCast(self), @ptrCast(param1));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QItemEditorCreatorBase_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qitemeditorfactory.html
pub const qitemeditorfactory = struct {

    /// New constructs a new QItemEditorFactory object.
    pub fn New() QtC.QItemEditorFactory {
        return qtc.QItemEditorFactory_new();
    }


    /// New2 constructs a new QItemEditorFactory object.
    pub fn New2(param1: ?*anyopaque) QtC.QItemEditorFactory {
        return qtc.QItemEditorFactory_new2(@ptrCast(param1));
    }


    pub fn CreateEditor(self: ?*anyopaque, userType: i32, parent: ?*anyopaque) QtC.QWidget {
        return qtc.QItemEditorFactory_CreateEditor(@ptrCast(self), userType, @ptrCast(parent));
    }

    /// Allows for overriding the related default method
    pub fn OnCreateEditor(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32, ?*anyopaque) callconv(.c) QtC.QWidget) void {
        qtc.QItemEditorFactory_OnCreateEditor(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseCreateEditor(self: ?*anyopaque, userType: i32, parent: ?*anyopaque) QtC.QWidget {
        return qtc.QItemEditorFactory_QBaseCreateEditor(@ptrCast(self), userType, @ptrCast(parent));
    }

    pub fn ValuePropertyName(self: ?*anyopaque, userType: i32, allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QItemEditorFactory_ValuePropertyName(@ptrCast(self), userType);
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qitemeditorfactory.ValuePropertyName: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    /// Allows for overriding the related default method
    pub fn OnValuePropertyName(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) [*:0]u8) void {
        qtc.QItemEditorFactory_OnValuePropertyName(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseValuePropertyName(self: ?*anyopaque, userType: i32, allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QItemEditorFactory_QBaseValuePropertyName(@ptrCast(self), userType);
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qitemeditorfactory.ValuePropertyName: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn RegisterEditor(self: ?*anyopaque, userType: i32, creator: ?*anyopaque) void {
        qtc.QItemEditorFactory_RegisterEditor(@ptrCast(self), userType, @ptrCast(creator));
    }

    pub fn DefaultFactory() QtC.QItemEditorFactory {
        return qtc.QItemEditorFactory_DefaultFactory();
    }

    pub fn SetDefaultFactory(factory: ?*anyopaque) void {
        qtc.QItemEditorFactory_SetDefaultFactory(@ptrCast(factory));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QItemEditorFactory_Delete(@ptrCast(self));
    }
};
